external-crates/move/crates/move-regex-borrow-graph/src/references.rs

```diff
@@ -0,0 +1,436 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    bail,
+    collections::{Path, Paths},
+    error,
+    regex::Regex,
+    Result,
+};
+use itertools::Either;
+use std::{
+    cmp,
+    collections::{BTreeMap, BTreeSet},
+    fmt::{self, Debug},
+};
+
+//**************************************************************************************************
+// Definitions
+//**************************************************************************************************
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+pub struct Ref(Ref_);
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd)]
+enum Ref_ {
+    Canonical(usize),
+    Fresh(usize),
+}
+
+#[derive(Clone)]
+struct LocRegex<Loc, Lbl> {
+    loc: Loc,
+    regex: Regex<Lbl>,
+}
+
+#[derive(Clone, Debug)]
+struct Edge<Loc, Lbl: Ord> {
+    abstract_size: usize,
+    regexes: BTreeSet<LocRegex<Loc, Lbl>>,
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct Node<Loc, Lbl: Ord> {
+    ref_: Ref,
+    is_mutable: bool,
+    abstract_size: usize,
+    self_epsilon: Regex<Lbl>,
+    successors: BTreeMap<Ref, Edge<Loc, Lbl>>,
```

Comment:
It seems odd to me that a node owns its edged -- this implicitly forms a tree that we then have to work with. Could we model a graph as a set of nodes and edges?

Reply:
I'm not sure I follow. There is no `Node` in the `Edge`, so it isn't really a tree, just an adjacency list/map